File: 0207.CourseSchedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        #graph = [[] for course in xrange(numCourses)]
        graph = {course: [] for course in xrange(numCourses)}
        inDegree = [0] * numCourses
        for pair in prerequisites:
            graph[pair[1]].append(pair[0])
            inDegree[pair[0]] += 1
        topo = []
        for course in xrange(numCourses):
            if inDegree[course] == 0:
                topo.append(course)
        i = 0
        while i < len(topo):
            for course in graph[topo[i]]:
                inDegree[course] -= 1
                if inDegree[course] == 0:
                    topo.append(course)
            i += 1
        return len(topo) == numCourses

File: 0000-0999/0207.CourseSchedule.go
func canFinish(numCourses int, prerequisites [][]int) bool {
    graph := make([][]int, numCourses)
    for _, p := range prerequisites {
        graph[p[0]] = append(graph[p[0]], p[1])
    }
    visit := make([]int, numCourses)
    for i := 0; i < numCourses; i++ {
        if !dfs(graph, visit, i) {
            return false
        }
    }
    return true
}

func dfs(graph [][]int, visit []int, i int) bool {
    if visit[i] == -1 {
        return false
    }
    if visit[i] == 1 {
        return true
    }
    visit[i] = -1
    for _, j := range graph[i] {
        if !dfs(graph, visit, j) {
            return false
        }
    }
    visit[i] = 1
    return true
}